PISTON-1191:Fixed a bug where tasks is creating multiple copies of the same object when tasks is running in multiple places across a distributed system

bradfordben opened the pull request:
After the removal of kz_globals from the tasks app the tasks app was creating N number of copies of an object where N was the number of tasks applications in the cluster.

Kazo 4.3 pr : https://github.com/2600hz/kazoo/pull/6701

bradfordben committed:
PISTON-1191:Fixed a bug where tasks is creating multiple copies of the same object when tasks is running in multiple places across a distributed system

jamesaimonetti commented:
@bradfordben what is a way to re-create the issue? I'd like to get a properly test setup to see this issue in action.

bradfordben commented:
@jamesaimonetti We found this issue when using the kt_users module that i have upstreamed but not gotten back to address your comments on (I do really apologize about this but the higher powers haven't given me time to address this but i will do it in the near future). So what we found was that when using that mod it would create N number of users where N was the number of kazoo tasks nodes in the system. I guess it would be reproducible for any of the tasks requests but as most of them just operate in the background and it would be hard to see that the task is being replicated on every tasks node and not just in one as it should be. Maybe kt_numbers might be of some use to recreate the issue

jamesaimonetti commented:
Looking at this more, I think the fix will be to move tasks.start_req processing to a named listener (usually something like `tasks_shared_listener`) where you have a fixed AMQP queue name. `tasks_listener` is "uniquely" named to recv copies of messages - in this case `lookup_req` should definitely be shared among all instances of tasks app, and stop/remove are fine too because any tasks instance that the job isn't running on will ignore the message.

So I think the proper fix is to use a shared listener for the start_req to ensure one tasks app starts the task.

icehess committed:
tinkering with rebar3 as alternative experimental build system (#36)

add rebar.config

icehess committed:
bump orb to use custom couchdb image (#35)

danielfinke committed:
gitignore `erlang_ls.config` per 2600hz/kazoo5@56c4cad (#33)

icehess committed:
update orb to consider os-release in cache keys (#37)

With upcoming changes, we need another variable to invalidate caches.

This adds /etc/os-release, uname -pmo and others together and uses them
as cache key in CI.

nkarunarathne committed:
Merge branch 'master' into tasks-creating-multiple-copies-of-objects-on-distributed-systems

bradfordben commented:
@jamesaimonetti Sorry i missed your message. With the team changes my focus has been elsewhere. Your suggestion makes sense to me. Unfortunately i dont have the time to do this as im focused elsewhere right now so feel free to close the PR.
